File: lib/javascript/src/math/Vector4D.js

ae.math.Vector4D = class Vector4D {
	
	backend:  ae.math.VectorBackend;
	xyz:      ae.math.Vector3D;
	readOnly: ae.math.Vector4D;
	
	x: number;
	y: number;
	z: number;
	w: number;
	
	constructor(
			backend:  ae.math.VectorBackend,
			_xyz:    ?ae.math.Vector3D) {
		
		if(backend instanceof ae.math.ReadOnlyBackend) {
			
			this.backend  = backend;
			this.xyz      = _xyz || new ae.math.Vector3D(backend);
			this.readOnly = this;
			
		} else {
			
			this.backend  = backend;
			this.xyz      = new ae.math.Vector3D(backend);
			this.readOnly = new Vector4D(
				new ae.math.ReadOnlyBackend(backend), this.xyz.readOnly);
		}
		
		this.x = 0;
		this.y = 0;
		this.z = 0;
		this.w = 0;
	}
	
	add(v: ae.math.Vevtor4D): ae.math.Vevtor4D {
		
		this.backend.x += v.backend.x;
		this.backend.y += v.backend.y;
		this.backend.z += v.backend.z;
		this.backend.w += v.backend.w;
		
		return this;
	}
	
	cloneConst(): ae.math.Vevtor4D {
		
		return new Vector4D(new ae.math.ReadOnlyBackend(
			new ae.math.StaticBackend(this.backend)));
	}
	
	cloneDeep(): ae.math.Vevtor4D {
		
		return new Vector4D(this.backend);
	}
	
	cloneStatic(): ae.math.Vevtor4D {
		
		return new Vector4D(new ae.math.StaticBackend(this.backend));
	}
	
	computeLength(): number {
		
		this.copyStaticValues();
		
		return Math.sqrt(
			this.x * this.x + this.y * this.y +
			this.z * this.z + this.w * this.w);
	}

	computeMean(): number {
		
		return (
			this.backend.x + this.backend.y +
			this.backend.z + this.backend.w) / 4;
	}

	copyStaticValues(): ae.math.Vevtor4D {
		
		this.x = this.backend.x;
		this.y = this.backend.y;
		this.z = this.backend.z;
		this.w = this.backend.w;
		
		return this;
	}

	static createConst(
			xOrGrey:  number,
			yOrW:    ?number,
			z:       ?number,
			w:       ?number): ae.math.Vector4D {
		
		w    = w    || yOrW || xOrGrey;
		z    = z    ||         xOrGrey;
		yOrW =         yOrW || xOrGrey;
		
		return new Vector4D(new ae.math.ReadOnlyBackend(
			new ae.math.StaticBackend(xOrGrey, yOrW, z, w)));
	}

	static createStatic(
			xOrGrey:  number,
			yOrW:    ?number,
			z:       ?number,
			w:       ?number): ae.math.Vector4D {
		
		w    = w    || yOrW || xOrGrey;
		z    = z    ||         xOrGrey;
		yOrW =         yOrW || xOrGrey;
		
		return new Vector4D(new ae.math.StaticBackend(xOrGrey, yOrW, z, w));
	}
	
	div(factor: number): ae.math.Vector4D {
		
		this.backend.x /= factor;
		this.backend.y /= factor;
		this.backend.z /= factor;
		this.backend.w /= factor;
		
		return this;
	}

	dot(v: ae.math.Vector4D): number {
		
		return (
			this.backend.x * v.backend.x + this.backend.y * v.backend.y +
			this.backend.z * v.backend.z + this.backend.w * v.backend.w);
	}
	
	getData(
			dst:    (ae.math.VectorBackend|ae.math.Vector4D|Array<number>),
			offset: number = 0):
			(ae.math.VectorBackend|ae.math.Vector4D|Array<number>) {
		
		if(dst instanceof Vector4D) {
			
			dst.setData(this.backend);
			
		} else if(dst instanceof ae.math.VectorBackend) {
			
			dst.x = this.backend.x;
			dst.y = this.backend.y;
			dst.z = this.backend.z;
			dst.w = this.backend.w;
			
		} else {
			
			dst[offset + 0] = this.backend.x;
			dst[offset + 1] = this.backend.y;
			dst[offset + 2] = this.backend.z;
			dst[offset + 3] = this.backend.w;
		}
		
		return dst;
	}
	
	mult(factorOrV: (number|ae.math.Vector4D)): ae.math.Vector4D {
		
		if(factorOrV instanceof Vector4D) {
			
			this.backend.x *= factorOrV.backend.x;
			this.backend.y *= factorOrV.backend.y;
			this.backend.z *= factorOrV.backend.z;
			this.backend.w *= factorOrV.backend.w;
			
		} else {
			
			this.backend.x *= factorOrV;
			this.backend.y *= factorOrV;
			this.backend.z *= factorOrV;
			this.backend.w *= factorOrV;
		}
		
		return this;
	}
	
	setData(
			xOrSrc: (ae.math.VectorBackend|ae.math.Vector4D|Array<number>|number),
			yOrOffset: number = 0,
			z:         number = 0,
			w:         number = 0): ae.math.Vector4D {
		
		if(xOrSrc instanceof Vector4D) {
			
			this.setData(this.backend);
			
		} else if(xOrSrc instanceof ae.math.VectorBackend) {
			
			this.backend.x = xOrSrc.x;
			this.backend.y = xOrSrc.y;
			this.backend.z = xOrSrc.z;
			this.backend.w = xOrSrc.w;
			
		} else if(typeof(xOrSrc) === "number") {
			
			this.backend.x = xOrSrc;
			this.backend.y = yOrOffset;
			this.backend.z = z;
			this.backend.w = w;
			
		} else {
			
			this.backend.x = xOrSrc[yOrOffset + 0];
			this.backend.y = xOrSrc[yOrOffset + 1];
			this.backend.z = xOrSrc[yOrOffset + 2];
			this.backend.w = xOrSrc[yOrOffset + 3];
		}
		
		return this;
	}
	
	sub(v: ae.math.Vector4D): ae.math.Vector4D {
		
		this.backend.x -= v.backend.x;
		this.backend.y -= v.backend.y;
		this.backend.z -= v.backend.z;
		this.backend.w -= v.backend.w;
		
		return this;
	}
	
	toZeroPoint(): ae.math.Vector4D {
		
		return this.setData(0, 0, 0, 1);
	}
	
	toZeroVector(): ae.math.Vector4D {
		
		return this.setData(0, 0, 0, 0);
	}
};

File: lib/java/src/ae/math/Vector4D.java
package ae.math;

import ae.util.OrganizedObject;

public final class Vector4D extends OrganizedObject<Vector4D> {

	public interface UnaryOperator {
		float applyToComponent(float x);
	}
	
	public static final Vector4D BLACK  = createConst(0,    1);
	public static final Vector4D GREY   = createConst(0.5f, 1);
	public static final Vector4D WHITE  = createConst(1,    1);
	public static final Vector4D RED    = createConst(1, 0, 0, 1);
	public static final Vector4D GREEN  = createConst(0, 1, 0, 1);
	public static final Vector4D BLUE   = createConst(0, 0, 1, 1);
	public static final Vector4D YELLOW = createConst(1, 1, 0, 1);
	public static final Vector4D PURPLE = createConst(1, 0, 1, 1);
	public static final Vector4D CYAN   = createConst(0, 1, 1, 1);
	
	public final VectorBackend backend;
	public final Vector3D      xyz;
	public final Vector4D      readOnly;
	
	public float x, y, z, w;
	
	private Vector4D(
			final ReadOnlyBackend backend,
			final Vector3D        xyz) {
		
		this.backend  = backend;
		this.xyz      = xyz;
		this.readOnly = this;
		
		backend.addListener((obj) -> _propagateChange());
	}
	
	public Vector4D(final ReadOnlyBackend backend) {
		
		this(backend, new Vector3D(backend));
	}
	
	public Vector4D(final VectorBackend backend) {
		
		this.backend  = backend;
		this.xyz      = new Vector3D(backend);
		this.readOnly =
			new Vector4D(new ReadOnlyBackend(backend), xyz.readOnly);
		
		backend.addListener((obj) -> _propagateChange());
	}
	
	public final Vector4D add(
			final Vector4D v) {
		
		backend.setX(backend.getX() + v.backend.getX());
		backend.setY(backend.getY() + v.backend.getY());
		backend.setZ(backend.getZ() + v.backend.getZ());
		backend.setW(backend.getW() + v.backend.getW());
		
		return this;
	}

	public final Vector4D applyUnaryOp(final UnaryOperator op) {
		
		backend.setX(op.applyToComponent(backend.getX()));
		backend.setY(op.applyToComponent(backend.getY()));
		backend.setZ(op.applyToComponent(backend.getZ()));
		backend.setW(op.applyToComponent(backend.getW()));
		
		return this;
	}
	
	public final Vector4D cloneConst() {
		
		return new Vector4D(new ReadOnlyBackend(new StaticBackend(backend)));
	}
	
	public final Vector4D cloneDeep() {
		
		return new Vector4D(backend);
	}
	
	public final Vector4D cloneStatic() {
		
		return new Vector4D(new StaticBackend(backend));
	}
	
	public final float computeLength() {
		
		copyStaticValues();
		
		return (float)Math.sqrt(x * x + y * y + z * z + w * w);
	}

	public final float computeMean() {
		
		return
			(backend.getX() + backend.getY() +
			 backend.getZ() + backend.getW()) / 4;
	}

	public final Vector4D copyStaticValues() {
		
		x = backend.getX();
		y = backend.getY();
		z = backend.getZ();
		w = backend.getW();
		
		return this;
	}

	public static final Vector4D createConst(
			final float grey) {
		
		return createConst(grey, grey, grey, grey);
	}
	
	public static final Vector4D createConst(
			final float grey,
			final float w) {
		
		return createConst(grey, grey, grey, w);
	}
	
	public static final Vector4D createConst(
			final float x,
			final float y,
			final float z,
			final float w) {
		
		return new Vector4D(new ReadOnlyBackend(new StaticBackend(x, y, z, w)));
	}

	public static final Vector4D createStatic(
			final float grey) {
		
		return createStatic(grey, grey, grey, grey);
	}
	
	public static final Vector4D createStatic(
			final float grey,
			final float w) {
		
		return createStatic(grey, grey, grey, w);
	}
	
	public static final Vector4D createStatic(
			final float x,
			final float y,
			final float z,
			final float w) {
		
		return new Vector4D(new StaticBackend(x, y, z, w));
	}
	
	public final Vector4D div(
			final float factor) {
		
		backend.setX(backend.getX() / factor);
		backend.setY(backend.getY() / factor);
		backend.setZ(backend.getZ() / factor);
		backend.setW(backend.getW() / factor);
		
		return this;
	}

	public final float dot(
			final Vector4D v) {
		
		return
			backend.getX() * v.backend.getX() +
			backend.getY() * v.backend.getY() +
			backend.getZ() * v.backend.getZ() +
			backend.getW() * v.backend.getW();
	}
	
	public final VectorBackend getData(
			final VectorBackend dst) {
		
		dst.setX(backend.getX());
		dst.setY(backend.getY());
		dst.setZ(backend.getZ());
		dst.setW(backend.getW());
		
		return dst;
	}
	
	public final Vector4D getData(
			final Vector4D dst) {
		
		return dst.setData(backend);
	}
	
	public final float[] getData(
			final float[] dst) {
		
		dst[0] = backend.getX();
		dst[1] = backend.getY();
		dst[2] = backend.getZ();
		dst[3] = backend.getW();
		
		return dst;
	}
	
	public final float[] getData(
			final float[] dst,
			final int     offset) {
		
		dst[offset    ] = backend.getX();
		dst[offset + 1] = backend.getY();
		dst[offset + 2] = backend.getZ();
		dst[offset + 3] = backend.getW();
		
		return dst;
	}

	public final Vector4D mult(
			final float factor) {
		
		backend.setX(backend.getX() * factor);
		backend.setY(backend.getY() * factor);
		backend.setZ(backend.getZ() * factor);
		backend.setW(backend.getW() * factor);
		
		return this;
	}
	
	public final Vector4D mult(
			final Vector4D v) {
		
		backend.setX(backend.getX() * v.backend.getX());
		backend.setY(backend.getY() * v.backend.getY());
		backend.setZ(backend.getZ() * v.backend.getZ());
		backend.setW(backend.getW() * v.backend.getW());
		
		return this;
	}
	
	public final Vector4D setData(
			final VectorBackend src) {
		
		backend.setX(src.getX());
		backend.setY(src.getY());
		backend.setZ(src.getZ());
		backend.setW(src.getW());
		
		return this;
	}
	
	public final Vector4D setData(
			final Vector4D src) {
		
		return setData(src.backend);
	}

	public final Vector4D setData(
			final float x,
			final float y,
			final float z,
			final float w) {
		
		backend.setX(x).setY(y).setZ(z).setW(w);
		
		return this;
	}
	
	public final Vector4D setData(
    		final float[] src) {
    	
    	return setData(src[0], src[1], src[2], src[3]);
    }
	
	public final Vector4D setData(
			final float[] src,
			final int     offset) {

		return setData(
			src[offset], src[offset + 1], src[offset + 2], src[offset + 3]);
	}

	public final Vector4D sub(
			final Vector4D v) {
		
		backend.setX(backend.getX() - v.backend.getX());
		backend.setY(backend.getY() - v.backend.getY());
		backend.setZ(backend.getZ() - v.backend.getZ());
		backend.setW(backend.getW() - v.backend.getW());
		
		return this;
	}
	
	public final Vector4D toZeroPoint() {
		
		return setData(0, 0, 0, 1);
	}
	
	public final Vector4D toZeroVector() {
		
		return setData(0, 0, 0, 0);
	}
}
